### ISSUE TITLE
Make build-system more user-friendly

### ISSUE THREAD

**Issue by dureuill** [trailofbits/remill#265]
Hello,

We've been evaluating remill and it looks great for our use, however when trying to integrate remill as a dependency into our CMake-based build system, we ran into some issues with its CMakeLists.txt.

The first issue is the usage of a LIBRARY_REPOSITORY_ROOT variable, which is not very CMake-friendly for external users:
* We would like to be able to specify the location for each dependency (gtest, glog, gflags) independently, and use the system one by default if available.
* When a LIBRARY_REPOSITORY_ROOT variable is set, the system LLVM is not looked up at all (only a direct lookup of the provided path).
* When a LIBRARY_REPOSITORY_ROOT variable is not set, the build system cannot find LLVM-related executables specified by LLVM_INSTALL_PREFIX (no hint in the `find_program` in the branch using a LIBRARY_REPOSITORY_ROOT)

The second issue is about remill's make install:
* The `make install` is not installing `libremill.a`, nor its headers. As a result, we have to manually copy the whole source directory to make the headers available publicly.
* What are the clang binaries generated by remill used for? Ideally we only need the library installed, can we safely remove those binaries?

Is there a way to require a specific LLVM version, beyond the LLVM_INSTALL_PREFIX? Also note that, even when using LLVM_INSTALL_PREFIX, if a system LLVM is installed, sometimes the `find_package` will prefer the system LLVM even with the hints (from [CMake's documentation of `find_package`](https://cmake.org/cmake/help/v3.9/command/find_package.html): "When multiple package configuration files are available whose version files claim compatibility with the version requested it is unspecified which one is chosen").

Also note the following about `find_package` scripts:
* Some `find_package` scripts that are not already present CMake (e.g. findXED) should be directly included in the repository
* The `find_package` scripts that are already in CMake (e.g. findGtest) should use the default scripts provided by CMake

We'd be willing to contribute PRs regarding these issues, to e.g. make it easier to build when not using a LIBRARY_REPOSITORY_ROOT or to install the library, if you can tell us more about your build system constraints.

**Comment by pgoodman** [comment]
Some comments/thoughts:

In terms of the first issue, one kludge to try right now that might work until we have a "real" solution is to make a `libraries` dir containing symlinks to all of the versions installed. I get that it's probably not friendly, but dependency management has been a pain for us in the past, hence why [cxx-common](https://github.com/trailofbits/cxx-common) exists.

In terms of the second point, @alessandrogario's Windows support branch improves things quite a bit, especially w.r.t. `make install`.

I am going to defer to @alessandrogario to what our constraints are or are not, and PRs that make things easier for your team and for others will definitely be welcome.

**Comment by alessandrogario** [comment]
Hey @dureuill,

it's great that you are considering remill for your project, and thanks for the good words :)

Let's talk about the issues you have mentioned:

## Installing the library
The new branch `alessandro/feature/windows-support` works better for the install target. 

Despite the name, it contains several fixes that are just not Windows related. Remill can now be installed on the system and can also be used to build other projects (like mcsema) without having the keep the full source repository around.

We are still in the process of reviewing it, so I would still encourage to use master though, and wait for it to be merged.

## Why we are installing the compiler executables
The reason we are installing them is (correct me if I'm wrong @pgoodman) that we need to make sure that the bitcode generated by the LLVM/Clang libraries that remill has been linked to is 100% compatible with the clang executable that the user will then have to use to manipulate it.

This is also important when deploying the tools on distributions that may not have access to the specific version of LLVM that we have used.

## How our CMake projects and dependency discovery works
Speaking about the project, we are using the standard way of importing libraries. The `LIBRARY_REPOSITORY_ROOT` parameter you mentioned is only used to include the [cxx-common cmake](https://github.com/trailofbits/cxx-common/blob/6099f902195949024d790a9d5e6b47903971df97/cmake_modules/repository.cmake) file that updates the standard [CMAKE_MODULE_PATH](https://cmake.org/cmake/help/latest/variable/CMAKE_MODULE_PATH.html) and [CMAKE_PREFIX_PATH](https://cmake.org/cmake/help/latest/variable/CMAKE_PREFIX_PATH.html) 
 variables used to locate dependencies. If you do not provide that parameter, the build system [will attempt to use system libraries](https://github.com/trailofbits/remill/blob/d50a2117fb5c66b5252a60d29a19e21c829068b2/CMakeLists.txt#L54).

Libraries that support CMake (like Google Flags or Google Log) are automatically picked up from the system if they have been installed. In case you want to use you own, you can just append its path to the CMAKE_PREFIX_PATH I have mentioned. Keep in mind that the system CMake files for LLVM are usually broken on Ubuntu (and possibly Debian too).

The remaining libraries that do not support CMake are not discoverable because they are missing a find_package handler, so you will have to provide one and [append the parent folder to CMAKE_MODULE_PATH](https://github.com/trailofbits/cxx-common/blob/6099f902195949024d790a9d5e6b47903971df97/cmake_modules/repository.cmake#L7). The cxx-common package stores them in the [cmake_modules folder](https://github.com/trailofbits/cxx-common/tree/6099f902195949024d790a9d5e6b47903971df97/cmake_modules).

## About the inclusion of the find_package scripts

I do not think that the missing find_package handlers should live inside this repository because remill is just a library; programs that make use of it may have different requirements or ways to discover dependencies.

Additionally, once you install remill on the system, you do not have them anymore unless you install them too (but that would not be something I consider standard). Another issue is that you can't (normally) link a static library against another one, unless you do it manually using `ar`, so we wouldn't be able to easily provided a consistent experience with and without cxx-common across all platforms.

## Why we are using cxx-common
1. Reproducible builds
2. Same development environment for everyone
3. No issues with outdated packages in the distro repository
4. Cross-platform compatibility (Windows, Linux, macOS). This is especially important for CI
5. Easy access to static libraries, so that we can run the compiled output everywhere
6. Access to the LLVM versions that we have tested to make sure that they are compatible with the code (it is sometimes not enough that the code compiles fine)
7. Be as generic as possible, without branding of any kind (so that it is usable by other developers as well)

## Building remill using system libraries
If you are not able to build and use remill without cxx-common and your own find_package handlers then it's a bug and we should fix it! I recently (a couple of months ago) built it on Arch Linux using system libraries (LLVM included). I'll test this again to make sure it works.

**Comment by pgoodman** [comment]
@dureuill can you evaluate @alessandrogario's windows build branch on your build targets and let us know how things work out?

**Comment by dureuill** [comment]
Hey @pgoodman, @alessandrogario,

We will evaluate the windows build branch with our build targets and let you know if this solves our issues.

Thank you for the very thorough answer regarding your build constraints. I'll forward that to my colleague who's the most comfortable with CMake, and we'll see how we can work from here.

Thank you again

**Comment by pgoodman** [comment]
@dureuill any updates?

**Comment by dureuill** [comment]
Hey @pgoodman, @alessandrogario,

Sorry for the delay, things have been a bit hectic over here these days ;-).

We evaluated the windows build branch with our build targets, I left a [comment](https://github.com/trailofbits/remill/pull/235#issuecomment-407818762) on the pull request.

Unfortunately we couldn't test if it solves our remill installation issue due to compilation issues (more details in my comment on the PR).

With this new branch it appears that we cannot select the LLVM version that we need to work with remill by specifying LLVM_INSTALL_PREFIX anymore. This is a major issue in our setup.

Thank you for the support regarding this issue :-).

**Comment by pgoodman** [comment]
@dureuill We've since merged the windows branch, and it's possible other things have changed. Is master working for you?

**Comment by dureuill** [comment]
OK, I'll have a look at it ASAP and report back wrt my issues.

**Comment by dureuill** [comment]
So, we tried the new master (on `c3bbeab43`, master has been moving quickly ;-), but it contains the windows branch) and save for some issues we were able to build it and launch the test.

For some of the issues, we found fixes, which I will submit as PRs.
However there's a couple of issues that we don't have fixes for, and for which we resorted to workarounds:
* The `make` target fails (error output at the end of this message). This seems related to `enable_language(BC)` in `tests/X86/CMakeLists.txt`. For the time being, commenting this line allows us to `make`, but I imagine that this doesn't constitute an acceptable fix?
* We ran into some difficulties for compiling the tests (I ran the `run-amd64-tests` target that @tathanhdinh gave me), because the mechanism used to set the `CMAKE_BC_COMPILER` and `CMAKE_BC_LINKER` defined in `cmake/BCCompiler.cmake` seems to differ from the mechanism used to set `CMAKE_LLVM_LINK`, `CMAKE_ASM_COMPILER`, ... in `cmake/utils.cmake`, and the former seems to ignore `LLVM_INSTALL_PREFIX` (to be precise, `CLANG_PATH` is set to the wrong value, even though `${LLVM_INSTALL_PREFIX}/bin` is passed in the `PATH` list of `find_program`). We can try to submit a PR but we don't know if our fix covers all cases.
* With these modifications, we can install and use remill in our build system, and we can run the tests, but they fail (16 failures, output at the end of this message, failures may be related to commenting out `enable_language(BC)`?).

I will try to proceed with the PR shortly :-).

Output of `make` target before we commented `enable_language(BC)`:
```
CMake Error: Could not find cmake module file: CMakeDetermineBCCompiler.cmake
CMake Error: Error required internal CMake variable not set, cmake may not be built correctly.
Missing variable is:
CMAKE_BC_COMPILER_ENV_VAR
CMake Error: Could not find cmake module file: /build/remill/src/remill-build/CMakeFiles/3.9.5/CMakeBCCompiler.cmake
```

Output of the `run-amd64-tests` executable:
```
----------] Global test environment tear-down
[==========] 1384 tests from 1 test case ran. (8656 ms total)
[  PASSED  ] 1368 tests.
[  FAILED  ] 16 tests, listed below:
[  FAILED  ] GeneralInstrTest/InstrTest.SemanticsMatchNative/401, where GetParam() = 0x1817d80
[  FAILED  ] GeneralInstrTest/InstrTest.SemanticsMatchNative/402, where GetParam() = 0x1817e00
[  FAILED  ] GeneralInstrTest/InstrTest.SemanticsMatchNative/423, where GetParam() = 0x1818880
[  FAILED  ] GeneralInstrTest/InstrTest.SemanticsMatchNative/424, where GetParam() = 0x1818900
[  FAILED  ] GeneralInstrTest/InstrTest.SemanticsMatchNative/643, where GetParam() = 0x181f680
[  FAILED  ] GeneralInstrTest/InstrTest.SemanticsMatchNative/644, where GetParam() = 0x181f700
[  FAILED  ] GeneralInstrTest/InstrTest.SemanticsMatchNative/647, where GetParam() = 0x181f880
[  FAILED  ] GeneralInstrTest/InstrTest.SemanticsMatchNative/648, where GetParam() = 0x181f900
[  FAILED  ] GeneralInstrTest/InstrTest.SemanticsMatchNative/651, where GetParam() = 0x181fa80
[  FAILED  ] GeneralInstrTest/InstrTest.SemanticsMatchNative/652, where GetParam() = 0x181fb00
[  FAILED  ] GeneralInstrTest/InstrTest.SemanticsMatchNative/734, where GetParam() = 0x1822400
[  FAILED  ] GeneralInstrTest/InstrTest.SemanticsMatchNative/1201, where GetParam() = 0x1830d80
[  FAILED  ] GeneralInstrTest/InstrTest.SemanticsMatchNative/1202, where GetParam() = 0x1830e00
[  FAILED  ] GeneralInstrTest/InstrTest.SemanticsMatchNative/1203, where GetParam() = 0x1830e80
[  FAILED  ] GeneralInstrTest/InstrTest.SemanticsMatchNative/1204, where GetParam() = 0x1830f00
[  FAILED  ] GeneralInstrTest/InstrTest.SemanticsMatchNative/1210, where GetParam() = 0x1831200
```

**Comment by pgoodman** [comment]
For your `grep` of the `run-amd64-tests` failures, can you grep for `States did not match`?

**Comment by alessandrogario** [comment]
Hey @dureuill,

thanks for trying out the new changes!

I have a couple of questions about the error you are experiencing; the `enable_language(BC)` statement should no longer be there according to the commit: https://github.com/trailofbits/remill/blob/c3bbeab4386213334ad17cab076f3171aa19d679/tests/X86/CMakeLists.txt

If the branch you are working on has been rebased on master maybe it has been caused by a merge commit (forcing a fast forward using `git rebase --ff-only` would help).

**Comment by dureuill** [comment]
@alessandrogario It looks to indeed be the case that my branch has some more lines than master. I just compared my branch to master, and indeed the `enable_language(BC)` is on my branch and not on master. Well that clears up why I was able to remove it safely.

However it means that I need to redo my tests starting from master proper. I will come back to you to update the results. I don't think it will have a major incidence on the PR I wanted to open.

@pgoodman I will check if I get the same results on master proper.

**Comment by dureuill** [comment]
I redid my tests on master. As I thought it doesn't really change the status I reported on 265#issuecomment-419142696, save for the first bullet point, which is nicely resolved :-).

I will try and submit some PRs shortly.

@pgoodman  I redid the tests on master proper, same result (16 failures). Attached are the results of the `grep` you suggested. I put it in a file, because it is 30080 lines long.
[remill_states_did_not_match.txt](https://github.com/trailofbits/remill/files/2361054/remill_states_did_not_match.txt)

**Comment by dureuill** [comment]
So I posted several PRs regarding this issue.
* #296, #297 have already been merged
* #298 is the PR corresponding to my issue with the `CMAKE_BC_COMPILER` variable. I'm unsure that this is what you want, so I can change my solution if needed, but we need this in some form to have `CMAKE_BC_COMPILER` correctly set in presence of a `LLVM_INSTALL_PREFIX`.
* I was about to post something similar to #294, but it has already been taken care of for us :+1: .

With all that, we're able to compile+run tests on master.
